### PR TITLE
feat: adds pin-get and pin-add lambda api

### PR DESCRIPTION
**Pull request by olizilla**
Add ipfs-cluster compatible api `BasicApi` that to allows to use pickup from web3.storage in place of our cluster. The db is focused on CIDs here rather than requestids as the pinning service api is.

Implements:
```
# get pin status for cid
GET  /pins/${cid}

# add pin
POST /pins/${cid}?origins=${multiaddr1},${multiaddr2}...
````

Also 
- adds basic auth protection to endpoints.
- responses are formatted to match what ipfs-cluster returns. These can be simplified once we are happy to switch over. Example responses are captured in [api/basic/examples/add-pin.json](https://github.com/olizilla/pickup/pull/19/files#diff-4cc47e0ac50f84789e66f101b7a94750c2604b9ad240fcebfee8e1b56f93065b) and [api/basic/examples/get-pin.json](https://github.com/olizilla/pickup/pull/19/files#diff-5ee99dd9545716fb3a9991f0eafcff40c68a6709b681c1be21c3f943ffe4e04c).

Fixes: https://github.com/olizilla/pickup/issues/16

## TODO:
- [x] return response objects should match ipfs-cluster responses
- [x] basic auth to match our cluster
- [x] add test for basic auth and http handlers

License: (Apache-2.0 AND MIT)
Signed-off-by: Oli Evans <oli@tableflip.io>

### CODE REVIEW

File: api/basic/add-pin.ts [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+ * We provide responses in Payload format v2.0
+ * see: https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html#http-api-develop-integrations-lambda.proxy-format
+ */
+export async function handler (event: APIGatewayProxyEventV2): Promise<Response> {
+  const {
+    TABLE_NAME: table = '',
+    BUCKET_NAME: bucket = '',
+    QUEUE_URL: queueUrl = '',
+    CLUSTER_BASIC_AUTH_TOKEN: token = '',
+    // set for testing
```
Comment (vasco-santos) [inline api/basic/add-pin.ts:40]: any other way of not setting this here for testing?

File: api/basic/add-pin.ts [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    DYNAMO_DB_ENDPOINT: dbEndpoint = undefined
+  } = process.env
+
+  if (event.headers.authorization !== `Basic ${token}`) {
+    return { statusCode: 401, body: { error: { reason: 'UNAUTHORIZED' } } }
+  }
+
+  const sqs = new SQSClient({ endpoint: sqsEndpoint })
+  const dynamo = new DynamoDBClient({ endpoint: dbEndpoint })
+  const cid = event.pathParameters?.cid ?? ''
```
Comment (vasco-santos) [inline api/basic/add-pin.ts:51]: I wouldn't default this to '' and adopt null instead. Then it will fail in `addPin`

File: api/basic/add-pin.ts [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+  const origins = event.queryStringParameters?.origins?.split(',') ?? []
+  try {
+    const res = await addPin({ cid, origins, bucket, sqs, queueUrl, dynamo, table })
+    return { ...res, body: JSON.stringify(res.body) }
+  } catch (error) {
+    console.log(error)
```
Comment (vasco-santos) [inline api/basic/add-pin.ts:57]: would be nice to adopt a logger with `warn`, `error`, `info`... Not needed in scope of this PR, but should exist for MVP

Comment (vasco-santos) [reply api/basic/add-pin.ts:57]: Also note that Errors from `addPin` won't be logged here. I would change them to throw an error instance with the reason and then in this catch would check if is instance of INTERNAL_SERVER_ERROR

File: api/basic/schema.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+export interface ClusterStatusResponse {
+  'cid': string
+  'name': ''
+  'allocations': []
+  'origins': []
+  'created': string
+  'metadata': null
+  'peer_map': {
+    '12D3KooWArSKMUUeLk3z2m5LKyb9wGyFL1BtWCT7Gq7Apoo77PUR': {
```
Comment (vasco-santos) [inline api/basic/schema.ts:20]: What if we change peerId? I would just set a Record here with string as key

File: api/basic/schema.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+export interface ClusterAddResponse {
+  replication_factor_min: -1
+  replication_factor_max: -1
+  name: ''
+  mode: 'recursive'
+  shard_size: 0
+  user_allocations: null
+  expire_at: '0001-01-01T00:00:00Z'
```
Comment (vasco-santos) [inline api/basic/schema.ts:40]: this shouldn't be hardcoded timestamp?

File: api/basic/get-pin.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+interface GetPinInput {
+  cid: string
+  dynamo: DynamoDBClient
+  table: string
+}
+
+/**
+ * AWS API Gateway handler for POST /pin/${cid}?&origins=${multiaddr},${multiaddr}
+ * Collect the params and delegate to addPin to do the work
```
Comment (vasco-santos) [inline api/basic/get-pin.ts:14]:
```suggestion
 * Collect the params and delegate to getPin to do the work
```

File: api/basic/get-pin.ts [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+import { ClusterStatusResponse, Pin, Response } from './schema.js'
+
+interface GetPinInput {
+  cid: string
+  dynamo: DynamoDBClient
+  table: string
+}
+
+/**
+ * AWS API Gateway handler for POST /pin/${cid}?&origins=${multiaddr},${multiaddr}
```
Comment (vasco-santos) [inline api/basic/get-pin.ts:13]:
```suggestion
 * AWS API Gateway handler for GET /pins/${cid}
```

**Review by vasco-santos — CHANGES_REQUESTED** [review]
Good re-work @olizilla . I don't have context on initial motivation, so left a couple of comments about previous handlers that are just to make that clear.

Other than that, left small pieces of feedback

File: stacks/BasicApiStack.ts [hunk 8]
```diff
@@ -1,18 +1,16 @@
 import { StackContext, Api, Table, Queue, Bucket } from '@serverless-stack/resources'
 
-export function ApiStack ({ stack }: StackContext): { queue: Queue, bucket: Bucket } {
+export function BasicApiStack ({ stack }: StackContext): { queue: Queue, bucket: Bucket } {
   const queue = new Queue(stack, 'Pin')
 
   const bucket = new Bucket(stack, 'Car')
 
-  const table = new Table(stack, 'PinStatusv6', {
+  const table = new Table(stack, 'BasicV2', {
```
[the remark this answers is not in the document]
Comment (olizilla) [reply stacks/BasicApiStack.ts:8]: You can only add additional indexes to a Table, you can't change it's definition, you have to recreate it. I have to manually go and delete the table when I want to try a new thing, so I've been incrementing the name while iterating on things.

File: api/basic/schema.ts (hunk 4 above)
Comment (olizilla) [reply api/basic/schema.ts:20]: yah, but it is a fake cluster peerID, there is no reason to change it. I agree it's suprising to see it hardcoded here, but it's only their to make the response the right shape. We might change the ipfs peer id / addr, but there is no reason to change the fake cluster peer id.

File: api/basic/schema.ts (hunk 5 above)
Comment (olizilla) [reply api/basic/schema.ts:40]: agree these look weird, but they currently capture the "fakeness" of our cluster responses. We don't support expiring pins here, so the value will always be that one. I can change the type defs if you feel strongly about it tho

File: api/test/basic.test.js [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+  t.context.dbEndpoint = dbEndpoint
+  t.context.dynamo = dynamo
+  t.context.table = table
+  t.context.sqsEndpoint = sqsEndpoint
+  t.context.sqs = sqs
+  t.context.createQueue = createQueue.bind(null, sqsContainer.getMappedPort(9324), sqs)
+})
+
+test.after.always(async t => {
+  t.context.containers.forEach(t => t.stop())
```
[the remark this answers is not in the document]
Comment (olizilla) [reply api/test/basic.test.js:50]: I had this as a best effort clean up, but don't want my test output to wait for it. I could probably just remove this and let testcontainer handle the clear up.

File: stacks/BasicApiStack.ts [hunk 10]
```diff
@@ -24,16 +22,14 @@ export function ApiStack ({ stack }: StackContext): { queue: Queue, bucket: Buck
...
+          CLUSTER_BASIC_AUTH_TOKEN: process.env.CLUSTER_BASIC_AUTH_TOKEN ?? ''
         }
       }
     },
     routes: {
-      'GET    /pins': 'functions/PinningService.handler',
-      'POST   /pins': 'functions/PinningService.handler',
-      'GET    /pins/{requestId}': 'functions/PinningService.handler',
-      'POST   /pins/{requestId}': 'functions/PinningService.handler',
-      'DELETE /pins/{requestId}': 'functions/PinningService.handler'
```
[the remark this answers is not in the document]
Comment (olizilla) [reply stacks/BasicApiStack.ts:36]: oh this is a confusing diff. I initially had both ApiStack and BasicApiStack in the repo, so we could switch between the two. I'm now leaning towards just dropping the old 'pinning service' ApiStack compeletly as we're no going to use it right now... we'll be using the cluster shaped api from our dotStorage workers.

I might publish the pinning service flavour as a separate repo as an example / runnable pinning service in a box kind of thing.

File: api/basic/add-pin.ts (hunk 3 above)
Comment (olizilla) [reply api/basic/add-pin.ts:57]: What's our preferred logger. Is it Pino?

File: api/basic/add-pin.ts (hunk 1 above)
Comment (olizilla) [reply api/basic/add-pin.ts:40]: I cant't change the lambda handler function signature to pass in an sqs or dynamo client, but to test it locally I need to use a client that points to the local endpoint. I don't think there is another way... but my hot brain might be missing something.

File: api/basic/add-pin.ts (hunk 3 above)
Comment (vasco-santos) [reply api/basic/add-pin.ts:57]: I remember seeing Pino usage on some of Elastic IPFS repos, would like to keep consistency. Perhaps worth asking the rest of the folks working on top of AWS what they are using

File: api/basic/schema.ts (hunk 4 above)
Comment (vasco-santos) [reply api/basic/schema.ts:20]: this is an Interface, we should be able to just create a new interface for PeerMapValue and do

```ts
`peer_map`: Record<string, PeerMapValue>
```

What do you think?

File: api/basic/schema.ts (hunk 5 above)
Comment (vasco-santos) [reply api/basic/schema.ts:40]: don't feel, so maybe just add a comment :)

File: api/basic/add-pin.ts (hunk 3 above)
Comment (olizilla) [reply api/basic/add-pin.ts:57]: Will do as another PR https://github.com/olizilla/pickup/issues/22

File: api/basic/add-pin.ts (hunk 2 above)
Comment (olizilla) [reply api/basic/add-pin.ts:51]: I'm trying out ts-standard on this project, and it won't let me do coercing if checks like if(someVar) unless someVar is actually a boolean, hence all this unidiomatic oddness. 

I can flip it to null, but both null and '' are checked for in addPin and the cid property is typed as string, so I'm using `''` as the as the only possible falsy value here. I agree it all feels a bit awkward tho.

File: api/basic/schema.ts (hunk 4 above)
Comment (olizilla) [reply api/basic/schema.ts]: Fixed as you suggested!

**Review by vasco-santos — APPROVED** [review]
LGTM 🎉

**Pull request merged**